sui_core/src/authority/authority_store.rs

```diff
@@ -753,61 +692,55 @@ impl<
-                //       full sequence, and the batching logic needs to deal with this.
-                write_batch = write_batch.insert_batch(
-                    &self.executed_sequence,
-                    std::iter::once((next_seq, transaction_digest)),
-                )?;
-            }
+        // NOTE: We just check here that locks exist, not that they are locked to a specific TX.  Why?
+        // 1. Lock existence prevents re-execution of old certs when objects have been upgraded
+        // 2. Not all validators lock, just 2f+1, so transaction should proceed regardless
+        //    (But the lock should exist which means previous transactions finished)
+        // 3. Equivocation possible (different TX) but as long as 2f+1 approves current TX its fine
+        if USE_LOCKS {
+            self.lock_service.locks_exist(owned_inputs.clone()).await?;
+        }
 
-            // Atomic write of all data other than locks
-            write_batch.write()?;
-            trace!("Finished writing batch");
+        if let Some(next_seq) = seq_opt {
+            // Now we are sure we are going to execute, add to the sequence
+            // number and insert into authority sequence.
+            //
+            // NOTE: it is possible that we commit to the database transactions
+            //       out of order with respect to their sequence number. It is also
+            //       possible for the authority to crash without committing the
+            //       full sequence, and the batching logic needs to deal with this.
+            write_batch = write_batch.insert_batch(
+                &self.executed_sequence,
+                std::iter::once((next_seq, transaction_digest)),
+            )?;
+        }
 
-            if USE_LOCKS {
-                // Initialize object locks for new objects.  After this point, transactions on new objects
-                // can run.  So it is critical this is done AFTER objects are done writing.
-                // TODO: add retries https://github.com/MystenLabs/sui/issues/1993
-                let new_locks_to_init: Vec<_> = written
-                    .iter()
-                    .filter_map(|(_, (object_ref, new_object))| {
-                        if new_object.is_owned() {
-                            Some(*object_ref)
-                        } else {
-                            None
-                        }
-                    })
-                    .collect();
-                self.lock_service
-                    .initialize_locks(new_locks_to_init)
-                    .await?;
-
-                // Remove the old lock - timing of this matters less
-                let locks_to_remove = owned_inputs;
-                self.lock_service.remove_locks(locks_to_remove).await?;
-            }
+        // Atomic write of all data other than locks
+        write_batch.write()?;
+        trace!("Finished writing batch");
 
-            // implicit: drop(_mutexes);
+        if USE_LOCKS {
```

Comment:
What happens if we crash here? (line 723) or in any place between writing the objects to the DB (and the sequence number), and updating the locks (make new ones / delete old ones) in lines 737 and 743? In fact I see we use `?` so we could fails on one of these and bail early, and leave with inconsistent state.

Reply:
Is this problem introduced with this PR or was it there before? (In any case it should be fixed)

Reply:
I suspect the problem appeared with the separation of locks from the DB write, before this PR.

Reply:
That's correct, this was just a whitespace change after changing the indent level.

Will take a look at this nonetheless

Reply:
Sun asked about this earlier too.  The whole idea of separating state is that we won't have absolute consistency any more, as things are separated into different stores.   You need one source of truth, which we can say could be the locks.  Basically, a new object could be written but is not final until the locks are done.   The thing is that as things become distributed or modularized, you won't ever have pure atomicity the way you do when everything is in one DB.

There are a couple of solutions:
1. To prevent inconsistent reads, you could read the locks first.  If the locks have not been cleared then the TX is not really done.  
2. We need to retry everything, especially the updating of the locks (deleting old ones and creating new ones).
3. We could make the lock update a single update instead of two (initialize new + delete old).
4. We could have some kind of rollback if the lock part fails.   This is kind of complicated though.
5. Instead of rollbacks, a simpler solution is to use a sequence number to determine what is the latest state which is "valid".  This would be a common way of solving this problem for distributed systems.  Basically, all object writes have a sequence number attached, which is also used for locks etc.  In order for a new state to be valid it must be beyond the 'good' sequence number which is updated only when everything succeeds.

I don't really like either 4 or 5, but I like 5 much better than 4.  

In reality, the chance that retries won't solve the problem and require some kind of rollback is exceedingly exceedingly small, especially for a local DB like RocksDB.   If retries failed, there is probably some serious problem and the process maybe should just panic.   When you restart, you clean up object state that doesn't make sense and stale locks etc.  IE if the outputs of a transaction are written but the locks aren't done, then we need to finish.

We can also just have "objects + locks" be a single atomic store, and split out certificates/effects.    Maybe that's the simplest solution of all.

Reply:
Let's add the discussions to https://github.com/MystenLabs/sui/issues/1990 so we can keep track